> **Tip**: Welcome to the Investigate a Dataset project! You will find tips in quoted sections like this to help organize your approach to your investigation. Before submitting your project, it will be a good idea to go back through your report and remove these sections to make the presentation of your work as tidy as possible. First things first, you might want to double-click this Markdown cell and change the title so that it reflects your dataset and investigation.

# Investigating the reasons behind No Show Medical Appointments

## Table of Contents
<ul>
<li><a href="#intro">Introduction</a></li>
<li><a href="#wrangling">Data Wrangling</a></li>
<li><a href="#eda">Exploratory Data Analysis</a></li>
<li><a href="#conclusions">Conclusions</a></li>
</ul>

<a id='intro'></a>
## Introduction

> **Tip**: In this section of the report, provide a brief introduction to the dataset you've selected for analysis. At the end of this section, describe the questions that you plan on exploring over the course of the report. Try to build your report around the analysis of at least one dependent variable and three independent variables.
>
> If you haven't yet selected and downloaded your data, make sure you do that first before coming back here. If you're not sure what questions to ask right now, then make sure you familiarize yourself with the variables and the dataset context for ideas of what to explore.

In [3]:
# Use this cell to set up import statements for all of the packages that you
#   plan to use.

# Remember to include a 'magic word' so that your visualizations are plotted
#   inline with the notebook. See this page for more:
#   http://ipython.readthedocs.io/en/stable/interactive/magics.html
import numpy as np,pandas as pd,matplotlib.pyplot as plt,seaborn as sns
%matplotlib inline

<a id='wrangling'></a>
## Data Wrangling

> **Tip**: In this section of the report, you will load in the data, check for cleanliness, and then trim and clean your dataset for analysis. Make sure that you document your steps carefully and justify your cleaning decisions.

### General Properties

In [32]:
# Load your data and print out a few lines. Perform operations to inspect data
#   types and look for instances of missing or possibly errant data.
df = pd.read_csv("noshow.csv")
df.head()

PatientId  AppointmentID Gender          ScheduledDay  \
0  2.987250e+13        5642903      F  2016-04-29T18:38:08Z   
1  5.589978e+14        5642503      M  2016-04-29T16:08:27Z   
2  4.262962e+12        5642549      F  2016-04-29T16:19:04Z   
3  8.679512e+11        5642828      F  2016-04-29T17:29:31Z   
4  8.841186e+12        5642494      F  2016-04-29T16:07:23Z   

         AppointmentDay  Age      Neighbourhood  Scholarship  Hipertension  \
0  2016-04-29T00:00:00Z   62    JARDIM DA PENHA            0             1   
1  2016-04-29T00:00:00Z   56    JARDIM DA PENHA            0             0   
2  2016-04-29T00:00:00Z   62      MATA DA PRAIA            0             0   
3  2016-04-29T00:00:00Z    8  PONTAL DE CAMBURI            0             0   
4  2016-04-29T00:00:00Z   56    JARDIM DA PENHA            0             1   

   Diabetes  Alcoholism  Handcap  SMS_received No-show  
0         0           0        0             0      No  
1         0           0        0             0      No  
2         0           0        0             0      No  
3         0           0        0             0      No  
4         1           0        0             0      No

In [41]:
df.nunique()

PatientId          62299
Gender                 2
ScheduledDay      103549
AppointmentDay        27
Age                  104
Neighbourhood         81
Scholarship            2
Hipertension           2
Diabetes               2
Alcoholism             2
Handcap                5
SMS_received           2
No-show                2
dtype: int64

Note here that Patient ID only has 62299 unique entries, compared to a total of 110527 actual appointments. Some patients had multiple appointments, so we cannot use PatientId as our index. AppointmentID is probably the most suitable.

In [42]:
df=pd.read_csv("noshow.csv",index_col="AppointmentID")
df.head()

PatientId Gender          ScheduledDay  \
AppointmentID                                              
5642903        2.987250e+13      F  2016-04-29T18:38:08Z   
5642503        5.589978e+14      M  2016-04-29T16:08:27Z   
5642549        4.262962e+12      F  2016-04-29T16:19:04Z   
5642828        8.679512e+11      F  2016-04-29T17:29:31Z   
5642494        8.841186e+12      F  2016-04-29T16:07:23Z   

                     AppointmentDay  Age      Neighbourhood  Scholarship  \
AppointmentID                                                              
5642903        2016-04-29T00:00:00Z   62    JARDIM DA PENHA            0   
5642503        2016-04-29T00:00:00Z   56    JARDIM DA PENHA            0   
5642549        2016-04-29T00:00:00Z   62      MATA DA PRAIA            0   
5642828        2016-04-29T00:00:00Z    8  PONTAL DE CAMBURI            0   
5642494        2016-04-29T00:00:00Z   56    JARDIM DA PENHA            0   

               Hipertension  Diabetes  Alcoholism  Handcap  SMS_received  \
AppointmentID                                                              
5642903                   1         0           0        0             0   
5642503                   0         0           0        0             0   
5642549                   0         0           0        0             0   
5642828                   0         0           0        0             0   
5642494                   1         1           0        0             0   

              No-show  
AppointmentID          
5642903            No  
5642503            No  
5642549            No  
5642828            No  
5642494            No

In [43]:
print(df.shape)
df.info()

(110527, 13)
<class 'pandas.core.frame.DataFrame'>
Int64Index: 110527 entries, 5642903 to 5629448
Data columns (total 13 columns):
PatientId         110527 non-null float64
Gender            110527 non-null object
ScheduledDay      110527 non-null object
AppointmentDay    110527 non-null object
Age               110527 non-null int64
Neighbourhood     110527 non-null object
Scholarship       110527 non-null int64
Hipertension      110527 non-null int64
Diabetes          110527 non-null int64
Alcoholism        110527 non-null int64
Handcap           110527 non-null int64
SMS_received      110527 non-null int64
No-show           110527 non-null object
dtypes: float64(1), int64(7), object(5)
memory usage: 11.8+ MB


In [44]:
df.dtypes

PatientId         float64
Gender             object
ScheduledDay       object
AppointmentDay     object
Age                 int64
Neighbourhood      object
Scholarship         int64
Hipertension        int64
Diabetes            int64
Alcoholism          int64
Handcap             int64
SMS_received        int64
No-show            object
dtype: object

No show column should be converted into 1/0 integers or boolean type. It should also be reversed to state whether a patient showed up or not, instead of the current meaning of whether a patient did or did not not show up, which is confusing.

In [47]:
type(df['PatientId'].iloc[0])

numpy.float64

PatientId is incorrectly categorized as a float, when it should be an integer

In [48]:
print(type(df["ScheduledDay"].iloc[0]))
print(type(df["AppointmentDay"].iloc[0]))

<class 'str'>
<class 'str'>


These two columns are string types but should be converted to datetime

In [49]:
df.describe()

PatientId            Age    Scholarship   Hipertension  \
count  1.105270e+05  110527.000000  110527.000000  110527.000000   
mean   1.474963e+14      37.088874       0.098266       0.197246   
std    2.560949e+14      23.110205       0.297675       0.397921   
min    3.921784e+04      -1.000000       0.000000       0.000000   
25%    4.172614e+12      18.000000       0.000000       0.000000   
50%    3.173184e+13      37.000000       0.000000       0.000000   
75%    9.439172e+13      55.000000       0.000000       0.000000   
max    9.999816e+14     115.000000       1.000000       1.000000   

            Diabetes     Alcoholism        Handcap   SMS_received  
count  110527.000000  110527.000000  110527.000000  110527.000000  
mean        0.071865       0.030400       0.022248       0.321026  
std         0.258265       0.171686       0.161543       0.466873  
min         0.000000       0.000000       0.000000       0.000000  
25%         0.000000       0.000000       0.000000       0.000000  
50%         0.000000       0.000000       0.000000       0.000000  
75%         0.000000       0.000000       0.000000       1.000000  
max         1.000000       1.000000       4.000000       1.000000

<ol>
    <li>The Age column shows a minimum age of -1, which is clearly an error/mistake.</li>
    <li>Typos in the header: "Hipertension" should really be "Hypertension", etc.</li>
</ol>


In [40]:
for i,v in enumerate(df.columns):
    print(i,v)

0 PatientId
1 Gender
2 ScheduledDay
3 AppointmentDay
4 Age
5 Neighbourhood
6 Scholarship
7 Hipertension
8 Diabetes
9 Alcoholism
10 Handcap
11 SMS_received
12 No-show


> **Tip**: You should _not_ perform too many operations in each cell. Create cells freely to explore your data. One option that you can take with this project is to do a lot of explorations in an initial notebook. These don't have to be organized, but make sure you use enough comments to understand the purpose of each code cell. Then, after you're done with your analysis, create a duplicate notebook where you will trim the excess and organize your steps so that you have a flowing, cohesive report.

> **Tip**: Make sure that you keep your reader informed on the steps that you are taking in your investigation. Follow every code cell, or every set of related code cells, with a markdown cell to describe to the reader what was found in the preceding cell(s). Try to make it so that the reader can then understand what they will be seeing in the following cell(s).

### Data Cleaning (Replace this with more specific notes!)

1. Remove duplicates
2. Correct data types of columns
3. Reverse category in No-show to a more intuitive understanding and convert to 1 or 0
4. Fix typos in headers

First, we need to check for duplicate data and clean it.

In [68]:
df=df.drop_duplicates()

In [69]:
df_undup.shape

(109909, 13)

In [70]:
# After discussing the structure of the data and any problems that need to be
#   cleaned, perform those cleaning steps in the second part of this section.
df.head()

PatientId Gender          ScheduledDay  \
AppointmentID                                              
5642903        2.987250e+13      F  2016-04-29T18:38:08Z   
5642503        5.589978e+14      M  2016-04-29T16:08:27Z   
5642549        4.262962e+12      F  2016-04-29T16:19:04Z   
5642828        8.679512e+11      F  2016-04-29T17:29:31Z   
5642494        8.841186e+12      F  2016-04-29T16:07:23Z   

                     AppointmentDay  Age      Neighbourhood  Scholarship  \
AppointmentID                                                              
5642903        2016-04-29T00:00:00Z   62    JARDIM DA PENHA            0   
5642503        2016-04-29T00:00:00Z   56    JARDIM DA PENHA            0   
5642549        2016-04-29T00:00:00Z   62      MATA DA PRAIA            0   
5642828        2016-04-29T00:00:00Z    8  PONTAL DE CAMBURI            0   
5642494        2016-04-29T00:00:00Z   56    JARDIM DA PENHA            0   

               Hipertension  Diabetes  Alcoholism  Handcap  SMS_received  \
AppointmentID                                                              
5642903                   1         0           0        0             0   
5642503                   0         0           0        0             0   
5642549                   0         0           0        0             0   
5642828                   0         0           0        0             0   
5642494                   1         1           0        0             0   

              No-show  
AppointmentID          
5642903            No  
5642503            No  
5642549            No  
5642828            No  
5642494            No

In [72]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 109909 entries, 5642903 to 5629448
Data columns (total 13 columns):
PatientId         109909 non-null float64
Gender            109909 non-null object
ScheduledDay      109909 non-null object
AppointmentDay    109909 non-null object
Age               109909 non-null int64
Neighbourhood     109909 non-null object
Scholarship       109909 non-null int64
Hipertension      109909 non-null int64
Diabetes          109909 non-null int64
Alcoholism        109909 non-null int64
Handcap           109909 non-null int64
SMS_received      109909 non-null int64
No-show           109909 non-null object
dtypes: float64(1), int64(7), object(5)
memory usage: 11.7+ MB


<a id='eda'></a>
## Exploratory Data Analysis

> **Tip**: Now that you've trimmed and cleaned your data, you're ready to move on to exploration. Compute statistics and create visualizations with the goal of addressing the research questions that you posed in the Introduction section. It is recommended that you be systematic with your approach. Look at one variable at a time, and then follow it up by looking at relationships between variables.

### Research Question 1 (Replace this header name!)

In [ ]:
# Use this, and more code cells, to explore your data. Don't forget to add
#   Markdown cells to document your observations and findings.


### Research Question 2  (Replace this header name!)

In [ ]:
# Continue to explore the data to address your additional research
#   questions. Add more headers as needed if you have more questions to
#   investigate.


<a id='conclusions'></a>
## Conclusions

> **Tip**: Finally, summarize your findings and the results that have been performed. Make sure that you are clear with regards to the limitations of your exploration. If you haven't done any statistical tests, do not imply any statistical conclusions. And make sure you avoid implying causation from correlation!

> **Tip**: Once you are satisfied with your work, you should save a copy of the report in HTML or PDF form via the **File** > **Download as** submenu. Before exporting your report, check over it to make sure that the flow of the report is complete. You should probably remove all of the "Tip" quotes like this one so that the presentation is as tidy as possible. Congratulations!